pep795.md

```diff
@@ -0,0 +1,1938 @@
+# Deep Immutability for Efficient Sharing and Concurrency Safety
+
+This PEP proposes that Python objects can be made immutable after
+which attempts to mutate them results in an exception. We will
+refer to the act of making an object immutable as "freezing it".
+There is no way to make an immutable object mutable again.
+However, one can make a mutable copy of an immutable object.
+
+The motivation for immutability is four-fold: 
+
+1. Immutability is a very useful concept in programming that
+   simplifies programming. In particular immutable objects are
+   safe to share between concurrent threads without any
+   synchronisation in the program.
+2. Immutable objects can be shared directly by reference
+   across subinterpreters. This is a big performance boost
+   for subinterpreters as Python currently requires objects
+   to be transferred across subinterpreters by pickling.
+   This also simplifies programming with subinterpreters.
+3. Since immutable objects are safe from data races, it is
+   possible to avoid some of the overhead surrounding field access
+   that is needed for correct reference counting in free-threaded
+   Python.
+4. Memory for immutable objects can be safely managed without leaks
+   without cycle detection, even when there are cycles in the 
+   immutable objects. This takes pressure off normal cycle 
+   detection in both subinterpreters and free-threading, and
+   is key to sharing objects between subinterpreters without
+   needing substantial changes to cycle detection.
+
+In this PEP, we strive to create a semantics for immutable
+objects that is consistent across both subinterpreters and
+free-threaded Python. This is a difficult problem as
+subinterpreters rely on isolation plus a per-interpreter GIL for
+correctness, whereas free-threading has no GIL.
+
+
+# Semantics of Immutability
```

Comment:
From just the introduction, I don't feel fully convinced that this is something Python needs. I think we should have another section before this showing at least one advantage you just listed above. Maybe a simple example that distributes work on a list across subinterpreters, and then say: "On the current version this requires the result to be pickeld. With immutability we can share the same array across multiple subinterprepers, not only saving performance from not pickeling, but also memory since we have a singular copy"

(This was me trying to channel my inner salesman xD)

Reply:
Again great. In the new 3-stage PEP I am listing direct sharing in subinterpreters as the 2nd motivation.